### PR TITLE
cluster: make dynamic updates to address-list provably thread-safe for LogicalHost, but not incur lock overhead for HostImpl.

### PR DESCRIPTION
**Pull request by jmarantz**
Commit Message: Resolves a race condition due to changing address-sets for LogicalHost during operation. The previous Host/HostDescription was hard to understand and prove correct from a thread safety perspective. In particular, it  returned a reference to a mutable structured via an API, and that is hard to protect from races. This PR resolves this by:

1. Returning a `shared_ptr<const vector<address>>` rather than a ref to a mutable vector from `addressList()`
2. Uses lock annotations to ensure member variables are safe during multi-threaded operations for LogicalHost
3. Uses `const` annotations to ensure member varaibles are  immutable and thus also threadsafe for HostImpl.
4. Adjusts class hierarchy to allow for maximum sharing of implementation between HostImpl and LogicalHost, but each one has its own member variables where necessary for the different thread safety models.

Additional Description:
WIthout the mutex, the new integraiton test fails under tsan 13/100 times. With the mutex it passes all the time. Note also that the mutex was there before this PR, but its use was not rigorously enforced by class structure or annotations. The fix was to change the API to return the address-list by shared_ptr rather than by reference.

Risk Level: medium -- this change to a central data-plane structure requires scrutiny. However the changes are mostly mechanical in order to enable the compiler to ensure thread safety for us.
Testing: //test/...
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
Fixes: #32850

### CODE REVIEW

**Comment by jmarantz** [comment]
I think the main outstanidng prolbem now is possibly coverage.

But also the new test (copied from a different PR) has some issues in it (statics that should be member vars, and is in the protoocl integration test when it should be a unit test on LogicalHost I think.

Otherwise I think this is mostly ready.

File: source/common/upstream/upstream_impl.h
Comment (adisuissa) [inline source/common/upstream/upstream_impl.h:336]: Perf suggestion: instead of calling `make_shared<>` here, create a staticexpr object of a SharedConstAddressVector type and return it.
This avoids allocating an empty vector (and shared-pointer) by multiple calls.

File: source/extensions/clusters/common/logical_host.cc
Comment (adisuissa) [inline source/extensions/clusters/common/logical_host.cc:37]: nit: const, or std::move later

File: source/extensions/clusters/common/logical_host.cc
Comment (adisuissa) [inline source/extensions/clusters/common/logical_host.cc:40]: The copy here is done while the lock is held.
I think that if the argument type for AddressVector will be a shared-ptr to const vector, that copy will not be needed. Can this be done instead?

File: source/extensions/clusters/common/logical_host.h
Comment (adisuissa) [inline source/extensions/clusters/common/logical_host.h:79]: Looking at `setHealthCheckerImpl`, `setOutlierDetector` and `setLastHcPassTime` both implementations seem to do the same. Can they just be implemented once at the base class, or am I missing something?

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this!
I like the idea of using mutexes to convince ourselves that the access to each data-member is guarded (although some reasoning is still needed for multi-members access).

File: source/common/upstream/upstream_impl.h
Comment (jmarantz) [reply source/common/upstream/upstream_impl.h:336]: Nice idea. But we can't use constexpr for shared_ptr: `'shared_ptr<const std::vector<std::shared_ptr<const Envoy::Network::Address::Instance>>>' is not literal because its destructor is not constexpr`

I don't think it would be a perf win here to do a lazy-init static because that requires a hidden cross-thread lock and creating an empty vector should be cheap (one alloc).

If we had a context in which to work with we could store it there. Maybe the ClusterInfo? But I don't want to make the PR bigger.

We could in a follow up have this return a std::optional or allow nullptr returns but aht would make the PR bigger as well.

File: source/extensions/clusters/common/logical_host.cc
Comment (jmarantz) [reply source/extensions/clusters/common/logical_host.cc:37]: done

File: source/extensions/clusters/common/logical_host.cc
Comment (jmarantz) [reply source/extensions/clusters/common/logical_host.cc:40]: done

Comment (jmarantz) [reply source/extensions/clusters/common/logical_host.cc:40]: oh...what's done is copying the address-list outside the lock.

I will leave a TODO to change the interface to take a shared_ptr which requires updating files outside the current PR.

File: source/extensions/clusters/common/logical_host.h
Comment (jmarantz) [reply source/extensions/clusters/common/logical_host.h:79]: You are right. The problem is the complex relationship where HostImpl and LogicalHostImpl have to inherit from 2 interfaces (Host and HostDescription) and 2 shared implementations (HostDescriptionImplBase and HostImplBase). I couldn't figure out how to share them or the 'resolve' function.

In the case of setHealthCheckerImpl et al, the problem, I think, is that HostDesciprtion does not have the setter API, but does have the getter API. This was pre-existing so I decided not to change it, but maybe this is worth it so that we don't have the new coverage issue.

I think the root

Comment (jmarantz) [reply source/extensions/clusters/common/logical_host.h:79]: I'm going to tackle this one as the alternative is to write new redundant tests to pass coverage. I think it might just add one file to the PR.

File: source/common/upstream/upstream_impl.cc
Comment (yanjunxiang-google) [inline source/common/upstream/upstream_impl.cc:466]: Here direct *addressList() maybe  risky, as addressList() could return a nullptr.

File: source/common/upstream/upstream_impl.h
Comment (yanjunxiang-google) [reply source/common/upstream/upstream_impl.h:335]: Can we return return address_list_ as it is? And in the caller code, if some places need to derefence it, then add the null check there, and return emptry vector {}?  
Just feel a little bit odd to create a shared_ptr() in this read accessor call.

File: source/common/upstream/upstream_impl.h
Comment (yanjunxiang-google) [inline source/common/upstream/upstream_impl.h:1361]: Do we really want to have an alias for existing std::vector<Network::Address::InstanceConstSharedPtr>, which trigger these kind of changes?

File: source/extensions/clusters/common/logical_host.cc
Comment (yanjunxiang-google) [inline source/extensions/clusters/common/logical_host.cc:65]: Instead of direct access the data members here, maybe you can add a new method to copy both address_ and address_list and have the method protected by lock?  With that, you can also move the address_ and address_list_ from protected into private section.

File: source/common/upstream/upstream_impl.h
Comment (yanjunxiang-google) [inline source/common/upstream/upstream_impl.h:503]: Yeah, right now, for several host types, the address_list is never populated and used, like static_cluster, strict_dns_cluster, original_dns_cluster,  ...

Maybe we can even further clean up the address_list from the HostDescriptionImpl, only keep address_ there?

File: test/common/upstream/upstream_impl_test.cc
Comment (yanjunxiang-google) [inline test/common/upstream/upstream_impl_test.cc:1551]: If we don't have address_list in the HostDescriiptionImpl, then we don't have to pass in it here.

File: source/common/upstream/upstream_impl.h
Comment (jmarantz) [reply source/common/upstream/upstream_impl.h:335]: I think it's safer to put this here rather than at (potentially) call-sites. But maybe if there's not many we could make them all check. I'll look at this.

File: source/common/upstream/upstream_impl.h
Comment (jmarantz) [reply source/common/upstream/upstream_impl.h:503]: I don't think so, no. I think the impl I have is tuned for that is we don't populate even an empty vector or shared_ptr; it just defaults to nullptr (see upstream_impl.cc:418).

File: source/common/upstream/upstream_impl.h
Comment (jmarantz) [reply source/common/upstream/upstream_impl.h:1361]: Yes I think it makes the code easier to read.

File: source/extensions/clusters/common/logical_host.cc
Comment (jmarantz) [reply source/extensions/clusters/common/logical_host.cc:65]: I could but it doesn't seem worth it; it's just 2 lines protected by a lock, and no other call-sites.

Comment (jmarantz) [reply source/extensions/clusters/common/logical_host.cc:65]: You did have the version that returned a pair before, and that's fine but I decided that was more verbose than just taking the lock and copying out the data.

File: source/common/upstream/upstream_impl.h
Comment (jmarantz) [reply source/common/upstream/upstream_impl.h:335]: I'm experimentally working on a commit to change the API to be called `addressListOrNull` and shift the burden to call-sites.

File: source/extensions/clusters/common/logical_host.cc
Comment (adisuissa) [inline source/extensions/clusters/common/logical_host.cc:42]: nit: maybe it is possible to std::move here.

Comment (adisuissa) [reply source/extensions/clusters/common/logical_host.cc:42]: Actually, I'm not sure what it buys here at the cost of readability, so leaving it up to you.

File: source/extensions/clusters/common/logical_host.h
Comment (adisuissa) [inline source/extensions/clusters/common/logical_host.h:35]: IMHO having a data member that is not accessed by the current class (even not the c'tor) may be an indication of a code-smell.
What are the tradeoffs if the `LogicalHostDescription` class is completely moved into `LogicalHost` while the latter ensures that all accesses are done using the mutex?

(If there are later uses of the `LogicalHostDescription` ignore this comment)

Comment (jmarantz) [reply source/extensions/clusters/common/logical_host.h]: Your sense of smell was spot on. Really there's no need to have LogicalHost and LogicalHOstDescription be separate; they weren't before; I think in the process of doing this refactor it seemed necessary but it clearly isn't. 

Done.

File: source/common/upstream/upstream_impl.cc
Comment (jmarantz) [reply source/common/upstream/upstream_impl.cc]: True, but I went the opposite direction; I made it protected member function of HostDescriptionImplBase and share it with LogicalHost, which had a similar block of code.

File: source/common/upstream/upstream_impl.h
Comment (jmarantz) [reply source/common/upstream/upstream_impl.h]: Yanjun this was a good call; we can move the checking to call-sites, there really was only one. I made it safe (hopefully) by renaming the function to addressListOrNull so it should be obvious.

File: test/integration/protocol_integration_test.cc
Comment (yanjunxiang-google) [inline test/integration/protocol_integration_test.cc:4927]: [This PR](https://github.com/envoyproxy/envoy/pull/32967) moved this test into a separate file, which can make the test run much faster:

File: test/integration/protocol_integration_test.cc
Comment (adisuissa) [inline test/integration/protocol_integration_test.cc:4961]: nit: move to the beginning of the test.

**Review by adisuissa — COMMENTED** [review]
Thanks, overall LGTM, modulo small nits.
I agree that it is safer to access the data-members this way. I'm not sure if there's a good way too measure the impact of using explicit locks on contention.

File: source/extensions/clusters/common/logical_host.h
Comment (jmarantz) [reply source/extensions/clusters/common/logical_host.h:92]: The root cause is, I think, because the HostDescription and Host abstractions are not quite right and I think there should not be an inheritance. Perhaps the abstractions were originally created before there was forward-proxy functionality.

Thus we need this delegate implementation for a Host which reflects flexibility and overrides for some of the fields and not other ones.

From a C++ perspective this manifests in a delegation tactic that uses a HostConstSharedPtr, so all the setters have to be no-ops.

I can add some comments.

File: test/integration/protocol_integration_test.cc
Comment (jmarantz) [reply test/integration/protocol_integration_test.cc:4961]: I'm going to drop this change once Yanjun's PR merges anyway.

**Review by jmarantz — COMMENTED** [review]
> I agree that it is safer to access the data-members this way. I'm not sure if there's a good way too measure the impact of using explicit locks on contention.

I think there are two changes at this level in the PR:
1. using thread annotations and `const` to ensure thread safety
2. swtching from *shared* (Reader) locks to *exclusive" locks

Number 1 is I think compellingly benficial.

For number 2 (exclusive locks), there's ample benchmarks, white-papers, and experience that shows that reader-locks are a performance anti-pattern when the critical sections are brief. They are better suited to situations where you need to hold the lock for a long time.

This is because reader-locks reqiure writing to the shared lock structure, and are more complex to manage. It's faster to use exclusive locks in most cases.

**Comment by yanjunxiang-google** [comment]
> > I agree that it is safer to access the data-members this way. I'm not sure if there's a good way too measure the impact of using explicit locks on contention.
> 
> I think there are two changes at this level in the PR:
> 
> 1. using thread annotations and `const` to ensure thread safety
> 2. swtching from _shared_ (Reader) locks to *exclusive" locks
> 
> Number 1 is I think compellingly benficial.
> 
> For number 2 (exclusive locks), there's ample benchmarks, white-papers, and experience that shows that reader-locks are a performance anti-pattern when the critical sections are brief. They are better suited to situations where you need to hold the lock for a long time.
> 
> This is because reader-locks reqiure writing to the shared lock structure, and are more complex to manage. It's faster to use exclusive locks in most cases.

I agree we should try 2).   The question is that should we interminge that within this PR.

**Comment by jmarantz** [comment]
I'm thinking just to push forward because it's the right thing to do for this usage model, and we are never going to come back to this.

File: test/integration/protocol_integration_test.cc
Comment (jmarantz) [reply test/integration/protocol_integration_test.cc]: I dropped this change as Yanjun moved this to its own file.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!
I think that it may warrant a release note (bug-fix), but leaving it up to the senior-maintainer to decide.

**Comment by jmarantz** [comment]
I think I should probaby go to @mattklein123 for senior approval for this one but Matt, feel free to redirect to others.

I'm curious how we should reason about the role of HostDescription vs Host, and I think this PR reveals a bit of confusion on my part there.

Note that we started poking around here because of crashes in production.

**Comment by yanjunxiang-google** [comment]
> I'm thinking just to push forward because it's the right thing to do for this usage model, and we are never going to come back to this.

Sounds good

**Review by mattklein123 — APPROVED** [review]
It's been years since I have looked at this code but this makes sense to me at a high level. Thanks for fixing!

**Pull request merged**
